### DISCUSSION TITLE
Issue on Need for Speed III: Hot Pursuit (1998)

### DISCUSSION THREAD

**Discussion by Goblan** [samuelgr/Xidi#60] [Q&A]
EDIT: Found the solution. Was planning to edit out all the thing but I'll leave the original message and the whole procedure here, in case someone else finds it useful.

ORIGINAL MESSAGE:
-----------------------
The game uses DirectInput. I can use my XInput controller with it and everything works fine. Everything but vibration/rumble/force feedback. The options related to it are in fact grayed out within the game's settings menu. I thought Xidi would fix the issue. I pasted dinput.dll file in the game's folder and now the game doesn't seem to recognize the gamepad buttons, sticks, or triggers. Vibration works but there's no way to play at all. The car seems to be accelerating and turning left the whole time. Can't do anything about it, can't change the camera, break, steer... Nothing. Am I doing something wrong? I would very much appreciate any help :)

I must add Xidi works a treat in other games such as Star Wars: Racer. Thank you for this beautiful piece of software.

SOLUTION:
------------
After copying Win32\dinput.dll to the game's main folder, create a Xidi.ini file in the same folder. It doesn't probably need to contain all of the text I'm about to paste. This is the general structure taken from the Readme section in this GitHub:

> [Mapper]
> Type                                = StandardGamepad
> Type.1                              = XInputNative
> Type.2                              = StandardGamepad
> Type.3                              = StandardGamepad
> Type.4                              = StandardGamepad
> 
> [Properties]
> MouseSpeedScalingFactorPercent      = 100
> UseBuiltInProperties                = yes
> DeadzonePercentStickLeft            = 0
> DeadzonePercentStickRight           = 0
> DeadzonePercentTriggerLT            = 0
> DeadzonePercentTriggerRT            = 0
> SaturationPercentStickLeft          = 100
> SaturationPercentStickRight         = 100
> SaturationPercentTriggerLT          = 100
> SaturationPercentTriggerRT          = 100
> 
> [Log]
> Enabled                             = no
> Level                               = 1
> 
> [Import]
> dinput.dll                          = C:\Windows\system32\dinput.dll
> dinput8.dll                         = C:\Windows\system32\dinput8.dll
> winmm.dll                           = C:\Windows\system32\winmm.dll
> 
> [CustomMapper]
> ; This section does not exist by default.
> 
> [Workarounds]
> ActiveVirtualControllerMask = 0x03
> PollReturnCode = 0

Since you're using an XInput controller, Mapper type shoud be changed to XInputNative, as in the quoted text, so that your previously saved control mappings in the game, if you made any, won't be all messed up. But the important part is at the bottom: the [workarounds](https://github.com/samuelgr/Xidi#workarounds).

There are 3 types in the Readme section. I find I had to use those two, otherwise it wouldn't work. 

Save the file and that's that.

Cheers!

**Comment by samuelgr** [comment]
Thanks for investigating. I'm guessing that for this game you would need your `Xidi.ini` file to look like this:

```ini
[Workarounds]
ActiveVirtualControllerMask = 0x03
PollReturnCode = 0
```

Do you need `ActiveVirtualControllerMask`? Getting rid of that line breaks the game?

**Reply by Goblan** [reply]
Sorry for the late response. If `ActiveVirtualControllerMask` line is not present, the game won't recognize the gamepad at all (and you won't be able to map any buttons either, let alone using rumble and stuff). I hope that helps. Thanks!

**Reply by Goblan** [reply]
After a few more tests I'v come to the conclusion that the vibration in this game via Xidi has no relation to what's going on in the game. I can feel variations in intensity, etc, but those just look random to me. For instance, if you crash into a traffic sign, vibration should increase immediately. It should change depending on the terrain, bumps, jumps, etc. That's the way it's always been, and Masahiko Morii's driver works perfectly in that regard. I wish Xidi worked just as good in this particular game. I've even thought that vibration with Xidi has some kind of delay. Honestly I'm not sure, but I think that's just not the case. To me it looks completely random and unrelated to the picture. I'd appreciate it if you could have a look. Thanks :)

**Reply by samuelgr** [reply]
I took a quick look at this game. Regarding the `Workarounds` section both lines are required (`ActiveVirtualControllerMask` and `PollReturnCode`) otherwise, yes, the game will not recognize inputs.

Force feedback generally does seem responsive (I can't reproduce the part about it seeming completely random, having tried with both an Xbox controller via Xidi and a Logitech RumblePad without Xidi). However, the game does use effects that Xidi doesn't currently support.

From the Xidi log:

```
[12/04/2023 16:35:44] [D] Creating effect with GUID Spring.
[12/04/2023 16:35:44] [I] Invoked Xidi::VirtualDirectInputDevice<0>::CreateEffect() on interface object 0 associated with Xidi virtual controller 1, result = 0x80040154.
```

Result code 0x80040154 indicates that the spring effect type isn't supported. I opened #73 for the additional force feedback types that aren't currently implemented.
